### ISSUE TITLE
Proposal: initiate rollback if --atomic is set and SIGINT is received

### ISSUE THREAD

**Issue by harryjsmith** [helm/helm#8040]
Sometimes, during a `helm upgrade --atomic`, I will notice that something has gone wrong which is going to cause the upgrade to fail. Rather than having to wait for the duration of the timeout period in order to have Helm perform the rollback, it would be useful if an initial ctrl+c/SIGINT triggered the rollback, rather than just killing the process dead and leaving things in a bad state. Perhaps a second ctrl+c/SIGINT sent in succession could additionally be used to indicate that the user would like to stop the process immediately.

**Comment by bacongobbler** [comment]
Please feel free to take a look into this and let us know what you find.

**Comment by lindhe** [comment]
@harryjsmith Yeah, I've thought about this too! Definitely how one wants the application to work. I've looked on it briefly. Considering that the wait seems to be happening [here](https://github.com/helm/helm/blob/master/pkg/action/install.go#L347), I assume we should listen for the the `SIGINT` signal in [this function](https://github.com/helm/helm/blob/3779d959667095e03308a2e49480748f7911b228/pkg/kube/client.go#L116).

That part looks fairly easy (except that I'm not familiar with signalling in Go). However, it might be tricky if we want different behavior from `--atomic` and `--wait` since `--atomic` implies wait and the existing logic really only considers the `--wait` case.

**Comment by lindhe** [comment]
Hi. I'd like this issue to not go stale, please github-actions bot...

**Comment by invidian** [comment]
not stale

**Comment by invidian** [comment]
not stale

**Comment by invidian** [comment]
Not stale..

**Comment by invidian** [comment]
Still an issue

**Comment by invidian** [comment]
So ideally all Helm operations should accept context as an input, so when the context is cancelled (e.g. by CLI receiving SIGINT), an rollback is triggered.

**Comment by invidian** [comment]
Not stale please.

**Comment by andy812** [comment]
I can't even imagine how they can fix this: when helm client receive SIGINT, its job is done. There is no tiller or something else in k8s cluster anymore, so what can trigger the rollback?
Maybe they could make some helm-operator what check bad state and rollback a release, idk.

**Comment by schibrikov** [comment]
@andy812 I guess SIGINT and SIGTERM could be catched to trigger a rollback. As I understand it's currently not the case and any cancellation just leads to a bad state.

**Comment by andy812** [comment]
@schibrikov yep, maybe they coud implement something like [trap](https://www.man7.org/linux/man-pages/man1/trap.1p.html) in Linux.

**Comment by bacongobbler** [comment]
https://github.com/helm/helm/pull/9180

**Comment by jorgeclaro** [comment]
Please note that this proposal should apply only for SIGINT signal when developing locally. Preferably with an additional flag to enable it. CI/CD pipelines using SIGTERM, should not start applying automated rollbacks without explicitly configured to do so.

On sandbox environments for example, automated rollbacks don't make any sence.

**Comment by philomory** [comment]
@jorgeclaro I mean, in those situations just don't use the `--atomic` flag? The whole point of the `--atomic` flag is to tell helm to roll back the release automatically if it fails; if you care about making it more difficult to debug broken releases, presumably you're not using the `--atomic` flag in the first place?

**Comment by jorgeclaro** [comment]
Got it! In that case I agree that when ` --atomic` flag is set the rollback should be executed on SIGINT signals. Probably it's not needed to wait for it to finish, but it should revert the release the previous state, at least.

**Comment by invidian** [comment]
Not stale please.

**Comment by bstrdsmkr** [comment]
Still not stale...

**Comment by joejulian** [comment]
I think this needs to be a contribution from the community. If someone would like to pick it up and open a PR, that would be fantastic.

Otherwise, when the next stale label comes around, I'm going to suggest that there's not enough interest and there's a sufficient workaround in just running `helm rollback` as a second step to close this issue. Sorry, as I know that folks can be pretty passionate about things that are important to them, but there's just not enough volunteers to tackle all the feature requests.

**Comment by jonnylangefeld** [comment]
I wanted to add to this conversation, the described behavior only happens upon `SIGSTP` (keyboard shortcut `C-z` or <kbd>Ctrl</kbd>-<kbd>z</kbd>) and not on `SIGINT` (keyboard shortcut `C-c` or <kbd>Ctrl</kbd>-<kbd>c</kbd>).

### Cancelling with <kbd>Ctrl</kbd>-<kbd>c</kbd>:

If you have a chart upgrading or installing with `--wait` or `--atomic` and hit <kbd>Ctrl</kbd>-<kbd>c</kbd> (`SIGINT`) while it's upgrading, everything works as expected and the error is the following:
```
^CRelease alpine has been cancelled.
Error: UPGRADE FAILED: release alpine failed, and has been rolled back due to atomic being set: context canceled
```
The result of the deployment is manifested in the release secret:
```
╰─ kubectl get secret sh.helm.release.v1.alpine.v7 -o json | jq -r '.data.release' | base64 -d | base64 -d | gunzip -c | jq '.info.status'
"failed"
```
If you try to upgrade the chart again at this point it will work.

### Cancelling with <kbd>Ctrl</kbd>-<kbd>z</kbd>:

However, if you doe the same but hit <kbd>Ctrl</kbd>-<kbd>z</kbd> (`SIGSTP`) while it's upgrading you'll see the following:
```
^Z
[1]  + 49813 suspended  helm upgrade -i --atomic alpine cmd/helm/testdata/testcharts/alpine-deploymen
```
This chart is stuck in the `pending-upgrade` status forever because there is no helm process running anymore getting it out of it:
```
╰─ kubectl get secret sh.helm.release.v1.alpine.v2 -o json | jq -r '.data.release' | base64 -d | base64 -d | gunzip -c | jq '.info.status'
"pending-upgrade"
```
If you try to upgrade the chart again at this point it will **not** work.

What's also interesting is that at this point helm even fails to list the installed charts:
```
╰─ helm ls
NAME	NAMESPACE	REVISION	UPDATED	STATUS	CHART	APP VERSION
```
even though the helm secrets are still there:
```
╰─ kg secret
NAME                           TYPE                 DATA   AGE
sh.helm.release.v1.alpine.v7   helm.sh/release.v1   1      4m28s
sh.helm.release.v1.alpine.v8   helm.sh/release.v1   1      4m24s
sh.helm.release.v1.alpine.v9   helm.sh/release.v1   1      19s
```
If you run a `helm rollback` at this point you can recover and after that `helm ls` will work again as well.

So whether or not your CI fails on this is really depending on which interruption signal it sends to the shell. Either `SIGINT` (works) or `SIGSTP` (doesn't work). Ours (Buildkite) unfortunately sends a `SIGSTP`

### Testing

All the testing for the above results was done using the following template:
<details>
<summary>yaml</summary>

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: "{{.Release.Name}}-{{.Values.Name}}"
  labels:
    # The "app.kubernetes.io/managed-by" label is used to track which tool
    # deployed a given chart. It is useful for admins who want to see what
    # releases a particular tool is responsible for.
    app.kubernetes.io/managed-by: {{.Release.Service | quote }}
    # The "app.kubernetes.io/instance" convention makes it easy to tie a release
    # to all of the Kubernetes resources that were created as part of that
    # release.
    app.kubernetes.io/instance: {{.Release.Name | quote }}
    app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
    # This makes it easy to audit chart usage.
    helm.sh/chart: "{{.Chart.Name}}-{{.Chart.Version}}"
    values: {{.Values.Name}}
spec:
  selector:
    matchLabels:
      app.kubernetes.io/instance: {{.Release.Name | quote }}
  template:
    metadata:
      labels:
        # The "app.kubernetes.io/managed-by" label is used to track which tool
        # deployed a given chart. It is useful for admins who want to see what
        # releases a particular tool is responsible for.
        app.kubernetes.io/managed-by: {{.Release.Service | quote }}
        # The "app.kubernetes.io/instance" convention makes it easy to tie a release
        # to all of the Kubernetes resources that were created as part of that
        # release.
        app.kubernetes.io/instance: {{.Release.Name | quote }}
        app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
        # This makes it easy to audit chart usage.
        helm.sh/chart: "{{.Chart.Name}}-{{.Chart.Version}}"
        values: {{.Values.Name}}
        ts: {{ now | date "20060102150405" | quote }}
    spec:
      # This shows how to use a simple value. This will look for a passed-in value
      # called restartPolicy. If it is not found, it will use the default value.
      # {{default "Never" .restartPolicy}} is a slightly optimized version of the
      # more conventional syntax: {{.restartPolicy | default "Never"}}
      restartPolicy: {{default "Always" .Values.restartPolicy}}
      terminationGracePeriodSeconds: 0
      initContainers:
      - name: wait
        image: "alpine:{{ .Chart.AppVersion }}"
        command: ["/bin/sleep","10"]
      containers:
      - name: waiter
        image: "alpine:{{ .Chart.AppVersion }}"
        command: ["/bin/sleep","9000"]
```
</details>

It upgrades the pod every single time due to the `ts: {{ now | date "20060102150405" | quote }}` label. It is designed to take 10 seconds for startup due to the init container. You can send the different interruption signals via <kbd>Ctrl</kbd>-<kbd>c</kbd> or <kbd>Ctrl</kbd>-<kbd>z</kbd> during the wait of upgrading your chart via 
```
helm upgrade -i --atomic alpine alpine-deployment
```

This was all tested with helm `v3.10.1`

# Conclusion

In my opinion helm's safeguard to catch a `SIGINT` signal should be extended to catch a `SIGSTP` signal. I don't think there is ever a case where anyone would want a half installed chart where they can only recover via a manual `helm rollback`, which is especially not feasible when a CI rolls everything out. We rollout >40 charts to over >70 clusters at once and no one would want to manually `helm rollback` everything if it's interrupted 😄.

**Comment by jkroepke** [comment]
> What's also interesting is that at this point helm even fails to list the installed charts

Its not a bug, its a feature, e.g. a default filter. In that case, use `helm ls --all`

**Comment by bstrdsmkr** [comment]
@joejulian I don't disagree on the community contribution, but I disagree on pretty much all the rest :grin:

This can be the root of many problems that aren't immediately obvious and can take quite some skill to track down. In our case, we're using Gitlab CI runners on Kubernetes with preemptible pipelines, very similar to above:

> this is really a problem when using helm in CI/CD pipelines. I use GitHub Actions with concurrency options specified to prevent concurrent deploys, so when new workflow is started the old one is terminated and the release is left in "pending" state, which leads second workflow to fail.

So when a pipeline is terminated, Gitlab asks kubernetes to kill the pod, k8s sends sigterm, waits for $gracefulTerminationSeconds, then sends sigkill. Both signals are send to docker (or other relevant runtime), passed on to whatever is PID1 in the container (usually a shell, but sometimes something like exec or tini, etc), then on down through the tree.

That's a lot of layers, any one of which failing causes this same symptom (release is left in failed state, subsequent runs can't recover). Assuming the user even knows the difference between sigkill and sigterm, I think the odds of giving up before figuring out which link in the chain failed are pretty high. If you close this issue, that's just going to further artificially suppress "demand" on this issue.

The rollback workaround is insufficient in this case (at least for Gitlab CI where you can't execute additional commands in response to a cancellation).

We are currently using a bit of a dirty hack to detect and rectify this situation on subsequent runs before attempting the next upgrade:
```yaml
before_script:
  # Cancelled pipelines don't propogate SIGTERM to the helm binary: https://gitlab.com/gitlab-org/gitlab-runner/-/issues/28162
  # This could leave releases in a pending state that helm can't recover from.
  # Find and delete any releases for this branch in failed states
  - >
    kubectl delete secret
    --ignore-not-found
    --namespace ${NAMESPACE}
    --selector ${RELEASENAME}
    --selector 'status in (pending-install, pending-upgrade, failed)'
script:
- helm upgrade --install --atomic ...
```

You'll note from the link in doc comment above that Gitlab has a way to go on proper signal handling, but I wanted to explain the context and give this workaround (ie, find any helm secrets that aren't completed and give them the ole' yeller) which should still work for this issue triggered in other ways.

> **Warning**
> Note that in my case, releases are uniquely named for the branch and duplicate jobs are automatically killed so it's always safe for me to rip out any pending-* installs. **__If you can't guarantee you don't have concurrent releases, adjust the selectors appropriately or you're gonna have a bad time.__**

tl;dr -- please leave the issue open until someone can contribute a fix so as not to further discourage contribution :grin:

**Comment by zarnovican** [comment]
if you came to this issue because cancelled Github Action helm leaves release in `pending-upgrade` state, read on..

```yaml
      - name: test1
        run: helm upgrade --namespace bzatest -i --atomic bzatest1 .
        working-directory: helm/
```

Github correctly sends SIGINT, SIGTERM "to the step's entry process" as [documented](https://docs.github.com/en/github-ae@latest/actions/managing-workflow-runs/canceling-a-workflow#steps-github-takes-to-cancel-a-workflow-run).

The problem is that `run:` action starts a bash/shell, which does not propagate the signal to its children. When cancelled task does not exit by itself in time, the whole process tree is eventually killed (by SIGKILL presumably).

## workaround: make helm the root process of Github action

Prefix the `helm` command with `exec` to replace top-level shell, eg:

```diff
-         run: helm upgrade --namespace bzatest -i --atomic bzatest1 .
+         run: exec helm upgrade --namespace bzatest -i --atomic bzatest1 .
```

Helm can correcly handle SIGINT (thank @jonnylangefeld for your comment). Here is an example output from Github run:

```
Release bzatest1 has been cancelled.
Error: UPGRADE FAILED: release bzatest1 failed, and has been rolled back due to atomic being set: context canceled
Error: The operation was canceled.
```

And the revision history:

```
REVISION	UPDATED                 	STATUS         	CHART         	APP VERSION	DESCRIPTION                                
..
37      	Thu Nov 24 12:39:27 2022	superseded     	bzatest1-0.1.0	latest     	Upgrade complete                           
38      	Thu Nov 24 12:41:11 2022	failed         	bzatest1-0.1.0	latest     	Upgrade "bzatest1" failed: context canceled
39      	Thu Nov 24 12:41:27 2022	deployed       	bzatest1-0.1.0	latest     	Rollback to 37                             
```

(tested with Helm 3.10.1)

**Comment by joejulian** [comment]
Nice find, @zarnovican!

**Comment by schibrikov** [comment]
Thanks a lot @zarnovican, it works!

There is an issue though - stdout is only visible on process exit, not in real-time. Do you know the possible reason for that and maybe some workaround? I don't see the logs, when e.g. helm waiting for pod or something and I find out the reason only on timeout.

**Comment by jkroepke** [comment]
@schibrikov  try `helm --debug upgrade`

**Comment by casey-robertson-paypal** [comment]
FYI this method is not working in a GH ARC runners environment in Kubernetes.
